### PR TITLE
Dependency reg dev

### PR DESCRIPTION
**Pull request by sterlingcrispin**
## Description of the change

This addresses #929 "Add dependency license concept to DependencyRegistry V0"

With this change there's a  `Bytes32Set _licenseTypes` which holds potential license types, and each `Dependency` struct now stores a `bytes32 licenseType ` of which license that dependency uses

When new dependencies are added, they must pass an argument that says what license the dependency is. This license must already exist in the set of  `_licenseTypes`.

New license types can be registered via `addLicenseType`

There's also some minor optimization of the existing contract, for example in `addDependency` rather than checking for `contains` before adding the dependency, 

```
        require(
            !_dependencyTypes.contains(_dependencyType),
            "Dependency type already exists"
        );
        _dependencyTypes.add(_dependencyType);
```

Instead, the function just tries to add the dependency, which returns false if the set already contained the dependency 

```
        require(
            // @dev the add function returns false if set already contains value
            _dependencyTypes.add(_dependencyType),
            "Dependency type already exists"
        );
```

The same pattern is now also used for adding and removing `_supportedCoreContracts`

@ryley-o @jakerockland

### CODE REVIEW

File: packages/contracts/contracts/DependencyRegistryV0.sol
Comment (ryley-o) [inline packages/contracts/contracts/DependencyRegistryV0.sol:682]: nit - would rather switch the order of these two statements, and only call `_licenseTypes.length()` once. We should follow that pattern in some of the other view functions in this contract as well (might make sense as follow-on work if you would like!)

File: packages/contracts/contracts/DependencyRegistryV0.sol
Comment (ryley-o) [inline packages/contracts/contracts/DependencyRegistryV0.sol:135]: I think it makes sense to only provide the ability to add new license types. Just wanted to double click and ensure you agree!

File: packages/contracts/contracts/DependencyRegistryV0.sol
Comment (ryley-o) [inline packages/contracts/contracts/DependencyRegistryV0.sol:682]: nit - would rather switch the order of these two statements, and only call `_licenseTypes.length()` once. We should follow that pattern in some of the other view functions in this contract as well (might make sense as follow-on work if you would like!)

File: packages/contracts/contracts/interfaces/v0.8.x/IDependencyRegistryV0.sol
Comment (ryley-o) [inline packages/contracts/contracts/interfaces/v0.8.x/IDependencyRegistryV0.sol:22]: nit: let's make this indexed to enable filtering to a specific license type

**Review by ryley-o — APPROVED** [review]
LGTM 🙏 couple of very minor nits/double-checks, and would like @jakerockland to give this another set of eyes once he returns, prior to merging into main.

File: packages/contracts/contracts/DependencyRegistryV0.sol
Comment (jakerockland) [reply packages/contracts/contracts/DependencyRegistryV0.sol:135]: ++

File: packages/contracts/contracts/DependencyRegistryV0.sol
Comment (jakerockland) [inline packages/contracts/contracts/DependencyRegistryV0.sol:138]:
```suggestion
            // @dev the add function returns false if set already contains value
            _licenseTypes.add(_licenseType),
```

**Review by jakerockland — APPROVED** [review]

File: packages/contracts/contracts/DependencyRegistryV0.sol
Comment (sterlingcrispin) [reply packages/contracts/contracts/DependencyRegistryV0.sol:133]: sounds good will close this comment

File: packages/contracts/contracts/DependencyRegistryV0.sol
Comment (sterlingcrispin) [reply packages/contracts/contracts/DependencyRegistryV0.sol]: Resolved here as an additional commit to this PR -> https://github.com/ArtBlocks/artblocks-contracts/pull/999/commits/c671723691f577f11b74f54f6f4890c325283a7a

**Pull request merged**
